examples/rpc_client.rs

```diff
@@ -12,36 +12,70 @@
  ********************************************************************************/
 mod common;
 
-use std::str::FromStr;
-use up_rust::{RpcClient, UMessageBuilder, UPayloadFormat, UUri};
-use up_transport_zenoh::UPClientZenoh;
+use async_trait::async_trait;
+use std::{
+    str::FromStr,
+    sync::{Arc, Mutex},
+};
+use tokio::time::{sleep, Duration};
+use up_rust::{
+    LocalUriProvider, UListener, UMessage, UMessageBuilder, UPayloadFormat, UTransport, UUri,
+};
+use up_transport_zenoh::UPTransportZenoh;
+
+// ResponseListener
+struct ResponseListener {
+    response: Mutex<Option<String>>,
+}
+impl ResponseListener {
+    fn new() -> Self {
+        ResponseListener {
+            response: Mutex::new(None),
+        }
+    }
+}
+#[async_trait]
+impl UListener for ResponseListener {
+    async fn on_receive(&self, msg: UMessage) {
+        let payload = msg.payload.unwrap();
+        let value = payload.into_iter().map(|c| c as char).collect::<String>();
+        let uri = msg.attributes.unwrap().source.unwrap().to_string();
+        let mut resp = self.response.lock().unwrap();
+        *resp = Some(value.clone());
+        println!("Receiving response {value} from {uri}");
+    }
+}
 
 #[tokio::main]
 async fn main() {
     // initiate logging
-    UPClientZenoh::try_init_log_from_env();
+    UPTransportZenoh::try_init_log_from_env();
 
     println!("uProtocol RPC client example");
-    let rpc_client = UPClientZenoh::new(common::get_zenoh_config(), String::from("rpc_client"))
+    let rpc_client = UPTransportZenoh::new(common::get_zenoh_config(), "//rpc_client/1/1/0")
         .await
         .unwrap();
 
     // create uuri
-    let src_uuri = UUri::from_str("//rpc_client/1/1/0").unwrap();
+    let src_uuri = rpc_client.get_source_uri();
     let sink_uuri = UUri::from_str("//rpc_server/1/1/1").unwrap();
 
-    // create uPayload
+    // register response callback
+    let resp_listener = Arc::new(ResponseListener::new());
+    rpc_client
+        .register_listener(&sink_uuri, Some(&src_uuri), resp_listener.clone())
+        .await
+        .unwrap();
+
+    // create uPayload and send request
     let data = String::from("GetCurrentTime");
     let umsg = UMessageBuilder::request(sink_uuri.clone(), src_uuri.clone(), 1000)
         .build_with_payload(data, UPayloadFormat::UPAYLOAD_FORMAT_TEXT)
         .unwrap();
+    println!("Sending request from {src_uuri} to {sink_uuri}");
+    rpc_client.send(umsg).await.unwrap();
 
-    // invoke RPC method
-    println!("Send request from {src_uuri} to {sink_uuri}");
-    let result = rpc_client.invoke_method(sink_uuri, umsg).await;
-
-    // process the result
-    let payload = result.unwrap().payload.unwrap();
-    let value = payload.into_iter().map(|c| c as char).collect::<String>();
-    println!("Receive {value}");
+    while resp_listener.response.lock().unwrap().is_none() {
```

Comment:
how about using [tokio::sync::Notify](https://docs.rs/tokio/latest/tokio/sync/struct.Notify.html) to notify the current thread about the message having been received?

Reply:
Good to know this API.